core/pkg/store/store.go

```diff
@@ -131,146 +210,180 @@ func (f *Store) String() (string, error) {
 }
 
 // GetAll returns a copy of the store's state (copy in order to be concurrency safe)
-func (f *Store) GetAll(_ context.Context) (map[string]model.Flag, model.Metadata, error) {
-	txn := f.db.Txn(false)
-
+func (s *Store) GetAll(ctx context.Context, selector *Selector, watcher chan Payload) (map[string]model.Flag, model.Metadata, error) {
+	txn := s.db.Txn(false)
 	flags := make(map[string]model.Flag)
-	it, err := txn.Get("flags", "id")
+	queryMeta := model.Metadata{}
+
+	var it memdb.ResultIterator
+	var err error
+	if selector != nil && !selector.isEmpty() {
+		queryMeta = selector.SelectorToMetadata()
+		indexId, constraints := selector.SelectorMapToQuery()
+		s.logger.Debug(fmt.Sprintf("getting all flags with query: %s, %v", indexId, constraints))
+		it, err = txn.Get("flags", indexId, constraints...)
+	} else {
+		// no selector, get all flags
+		it, err = txn.Get(flagsTable, idIndex)
+	}
 
 	if err != nil {
-		return flags, model.Metadata{}, err
+		s.logger.Error(fmt.Sprintf("query error: %v", err))
+		return flags, queryMeta, err
 	}
 
-	for obj := it.Next(); obj != nil; obj = it.Next() {
-		flag := obj.(model.Flag)
+	for raw := it.Next(); raw != nil; raw = it.Next() {
+		flag := raw.(model.Flag)
+		if existing, ok := flags[flag.Key]; ok {
+			if flag.Priority < existing.Priority {
+				s.logger.Debug(fmt.Sprintf("discarding duplicate flag %s from lower priority source %s in favor of flag from source %s", flag.Key, s.sources[flag.Priority], s.sources[existing.Priority]))
+				continue // we already have a higher priority flag
+			} else {
+				s.logger.Debug(fmt.Sprintf("overwriting duplicate flag %s from lower priority source %s in favor of flag from source %s", flag.Key, s.sources[existing.Priority], s.sources[flag.Priority]))
+			}
+		}
 		flags[flag.Key] = flag
 	}
 
-	return flags, f.getMetadata(), nil
+	if watcher != nil {
+
+		// a "one-time" watcher that will be notified of changes to the query
+		changes := it.WatchCh()
+
+		go func() {
+			select {
+			case <-changes:
+				s.logger.Debug("flags store has changed, notifying watchers")
+
+				// recursively get all flags again in bytes new goroutine to keep the watcher responsive
+				// as long as we do this in bytes new goroutine, we don't risk stack overflow
+				bytes, _, err := s.GetAll(ctx, selector, watcher)
+				if err != nil {
+					s.logger.Error(fmt.Sprintf("error getting flags in watcher: %v", err))
+					break
+				}
+				watcher <- Payload{
```

Comment:
`s.GetAll` will start another goroutine, which may close the `watcher` channel if the context is cancelled, and if it happens before L265 is executed, the program will panic